src/descriptor/checksum.rs

```diff
@@ -2,177 +2,180 @@
+        if add == DescriptorChecksum::CHECKSUM_LENGTH {
+            add += 1; // +1 for the # character
         }
-        chars
-    }
+        let (min, max) = self.iter.size_hint();
 
-    /// Obtains the checksum of all the data thus-far fed to the engine.
-    pub fn checksum(&mut self) -> String {
-        String::from_iter(self.checksum_chars().iter().copied())
+        (min + add, max.map(|max| max + add))
     }
 }
 
-/// A wrapper around a `fmt::Formatter` which provides checksumming ability.
-pub struct Formatter<'f, 'a> {
-    fmt: &'f mut fmt::Formatter<'a>,
-    eng: Engine,
-}
+/// The Output Script Descriptor checksum algorithm, defined in [BIP-380].
+///
+/// [BIP-380]: <https://github.com/bitcoin/bips/blob/master/bip-0380.mediawiki>
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+enum DescriptorChecksum {}
 
-impl<'f, 'a> Formatter<'f, 'a> {
-    /// Contructs a new `Formatter`, wrapping a given `fmt::Formatter`.
-    pub fn new(f: &'f mut fmt::Formatter<'a>) -> Self { Formatter { fmt: f, eng: Engine::new() } }
+/// Generator coefficients, taken from BIP-380.
+#[rustfmt::skip]
+const GEN: [u64; 5] = [0xf5dee51989, 0xa9fdca3312, 0x1bab10e32d, 0x3706b1677a, 0x644d626ffd];
 
-    /// Writes the checksum into the underlying `fmt::Formatter`.
-    pub fn write_checksum(&mut self) -> fmt::Result {
-        use fmt::Write;
-        self.fmt.write_char('#')?;
-        for ch in self.eng.checksum_chars().iter().copied() {
-            self.fmt.write_char(ch)?;
-        }
-        Ok(())
-    }
+const CHECKSUM_LENGTH: usize = 8;
+
+impl Checksum for DescriptorChecksum {
+    type MidstateRepr = u64; // We need 40 bits (8 * 5).
+    const CHECKSUM_LENGTH: usize = CHECKSUM_LENGTH;
+    const GENERATOR_SH: [u64; 5] = GEN;
+    const TARGET_RESIDUE: u64 = 1;
+}
 
-    /// Writes the checksum into the underlying `fmt::Formatter`, unless it has "alternate" display on.
-    pub fn write_checksum_if_not_alt(&mut self) -> fmt::Result {
-        if !self.fmt.alternate() {
-            self.write_checksum()?;
+/// Helper function for `FromStr` for various descriptor types.
+///
+/// Checks and verifies the checksum if it is present and returns the descriptor
+/// string without the checksum.
+pub(super) fn verify_checksum(s: &str) -> Result<&str, Error> {
+    for ch in s.as_bytes() {
+        if *ch < 20 || *ch > 127 {
+            return Err(Error::Unprintable(*ch));
         }
-        Ok(())
     }
-}
 
-impl<'f, 'a> fmt::Write for Formatter<'f, 'a> {
-    fn write_str(&mut self, s: &str) -> fmt::Result {
-        self.fmt.write_str(s)?;
-        self.eng.input(s).map_err(|_| fmt::Error)
+    let mut parts = s.splitn(2, '#');
+    let desc_str = parts.next().unwrap();
+
+    if let Some(_) = parts.next() {
+        let checksummed = Checksummed::new(desc_str)?;
+        if !checksummed.eq(s.chars()) {
+            return Err(Error::BadDescriptor("invalid checksum".to_owned()));
+        }
```

Comment:
In 22ee854c61b13d15fb7d0b11daf2a87604f45ba7:

It would be faster to feed the checksum into the checksum iterator and compare the residue against the target. Then in the happy case we only need to iterate once.

In the unhappy case, we should probably return a better error variant that has the expected and received checksum in it. It's fine if this is slow and allocating.

Reply:
Sure thing.